### PR TITLE
Force remount LinkControl when moving between links within same richtext block

### PR DESCRIPTION
**Pull request by getdave**
## Description

In https://github.com/WordPress/gutenberg/pull/32320 we learnt that the Link UI could be made to display values that were out of the sync with its underlying data.

We should still pursue https://github.com/WordPress/gutenberg/pull/32320 as a means of solving this issues at an underlying level. 

However, a contributing problem is that `<LinkControl>` **isn't being remounted** when it's used within the rich text package. This is because it's within a `<Popover />`, and because `<Popover>` isn't unmounted, as you click between links the `<LinkControl>` is also **_not_ unmounted**. This allows values to stay in state (and get shared) between what should be considered two entirely separate instances of `<LinkCongrol>`. This allows these bugs to manifest more easily.

I think it's reasonable to expect `<LinkControl />` to _remount_ as you click between links. Indeed if you are in edit mode on link 1 and then you click away on to link 2, then you should expect that you will no longer be in edit mode on link 2.

This PR forces that to happen via the addition of a `key` prop which is set to ~the stringified HTML of the anchor ref to which the link UI is "attached". This is not _guaranteed_ to be unique but 99.9% of the time it will be because you rarely make the same link within the same paragraph block with the exact same attributes~ a unique ID based key in the form `link-control-instance-${id}`. A unique ID is "assigned" to each LinkControl based on a map of anchorRef -> uniqueIDs. This map is stored as a `WeakMap` to avoid memory leaks. Hat tip to @kevin940726 for the inspiration.

Please note that `useInstanceId` could not be used because the component does not remount and thus the instance id generated by that hook remains identical between links. 

## How has this been tested?

1. Create new post.
2. Create single paragraph block and add some text.
3. Create x3 links:
   - `wordpress.org`
   - `wordpress.org` (yes that's correct I want you to add the same link!)
   - `wordpress.org/gutenberg`
4. Click on one of the links and click `Edit` to go into "edit mode".
5. Now click on any of the other links and check that you are forced to leave edit mode.

Bottom line - if you leave the link you are currently on you should have to click edit again.

## Screenshots 

https://user-images.githubusercontent.com/444434/140541582-4a661279-4e38-463e-be17-dafba262b7eb.mov

## Types of changes
 Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by getdave** [comment]
cc @ralucaStan I'd appreciate your 👀 on this one as well 🙏

**Comment by javierarce** [comment]
This worked as you described 🚀

**Comment by ralucaStan** [comment]
> cc @ralucaStan I'd appreciate your 👀 on this one as well 🙏

@getdave this looks and tests 👌 .

#32320 still solves the issue at URLInput's level, so that the issue won't happen outside of this particular Link UI. So I also think these two PRs work well together.

File: packages/format-library/src/link/inline.js [hunk 1]
```diff
@@ -222,6 +224,7 @@ function InlineLinkUI( {
 			position="bottom center"
 		>
 			<LinkControl
+				key={ linkControlRemountKey }
```
Comment (kevin940726) [inline packages/format-library/src/link/inline.js]: Comparing full HTMLs could be a little bit costly (in theory). Instead, I think we can use `useInstanceId` here to generate the key.

```js
const linkControlKey = useInstanceId( anchorRef );
```

It's using `WeakMap` under the hood, so shouldn't have memory leaks when the anchor gets destroyed.

Comment (getdave) [reply packages/format-library/src/link/inline.js]: @kevin940726 I did experiment with that. The issue is that the instance id appears to be reset which means it's not unique when I move between links



https://user-images.githubusercontent.com/444434/140468619-9d07a138-98c8-4c2d-a98e-b1fd370a092f.mov

Any idea what's going on there?

Comment (kevin940726) [reply packages/format-library/src/link/inline.js]: Oops, turns out that `useInstanceId` isn't what I thought it was. What I was thinking is rather like a `useUniqueId` or `useInstanceKey` or something like that.

```js
const weakMap = new WeakMap();
let id = -1;

function getKey( id, prefix ) {
  return prefix ? `${ prefix }-${ id }` : `${ id }`;
}

function useInstanceKey( instance, prefix ) {
  if ( weakMap.has( instance ) ) {
    return getKey( weakMap.get( instance ), prefix );
  }

  id += 1;
  weakMap.set( instance, id );

  return getKey( id, prefix );
}
```

TBH, I'm not really sure what `useInstanceId` is anymore.

Comment (getdave) [reply packages/format-library/src/link/inline.js]: Yeh I guess we can implement our own unique ID system here as it's a bit of an edge case.

Comment (getdave) [reply packages/format-library/src/link/inline.js]: @kevin940726 I've used your code as inspiration for a custom unique key generator. This seems to work. What do you think?

**Review by kevin940726 — APPROVED** [review]
LGTM 👍 .

Though I think the new hook worth belonging in the `@wordpress/compose` package but that's for another PR.

**Comment by andrewserong** [comment]
Cherry picked into the Gutenberg 11.9 release in: https://github.com/WordPress/gutenberg/commit/3fe22360dca325ec0e69edd288c4e03a5ffb35db
